Layout: Expand editor canvas as flex region

aduth opened the pull request:
Closes #2322

This pull request seeks to use flex styling to ensure the visual editor canvas occupies the full height of the screen, thereby enabling users to clear the selected block by clicking below the editor on a new post.

One noticeable difference in this implementation is that, when meta boxes are present, they will be pushed to the bottom of the viewport rather than appearing immediately below the post content.

Before|After
---|---
![Before](https://user-images.githubusercontent.com/1779930/37009484-2356c1fa-20b5-11e8-9aec-111b7fead7af.png)|![After](https://user-images.githubusercontent.com/1779930/37009479-1c1b8d30-20b5-11e8-9326-34f62fdf20fa.png)

__Testing instructions:__

Verify that with and without meta boxes, and with and without long content, that...

- Meta boxes appear at the bottom of the screen / post
- If there is any space below the post content (e.g. new post, it can be used to deselect the currently selected block)

aduth committed:
Layout: Expand editor canvas as flex region

jasmussen commented:
This seems to _vastly_ enhance the Gutenberg editing experience. 

Here's the canvas change in GIF form:

![flexcanvas](https://user-images.githubusercontent.com/1204802/37023615-cf88b570-2126-11e8-8061-f5aaae75075d.gif)

As is shown here, the issue with the metaboxes being at the bottom is primarily an issue for the "empty page". This tradeoff seems definitely worth making, because the ability to deselect early blocks by just clicking below the block in the large visible space, is such an improvement to the past behavior. It almost feels like the gloves coming off. 

However Riad touches on a good point, we did receive pushback back when metaboxes were in what looked like a detached panel. So just for due diligence, I'm CC'ing the leads to make a call: @karmatosed @mtias. 

Personally, I would give the go-ahead on this one for one simple reason: the disconnect between metaboxes at the bottom and an empty canvas above happens **only when those metaboxes are collapsed**:

<img width="1465" alt="screen shot 2018-03-06 at 10 13 31" src="https://user-images.githubusercontent.com/1204802/37023822-65bc1a50-2127-11e8-8ec5-7535296bf07e.png">

Since metaboxes maintain their state of being opened across sessions, on balance this seems like it should not be a blocker for enhancing the deselectability of blocks. 

👍👍 from me, but good to get a call from the leads.

aduth commented:
>Personally, I would give the go-ahead on this one for one simple reason: the disconnect between metaboxes at the bottom and an empty canvas above happens only when those metaboxes are collapsed:

Yes, my original example was maybe not the best to show a single collapsed meta box, which would appear at the bottom of the screen.

But the natural behavior here is that meta boxes will expand upward if there is space, and share the available canvas with content, with a preference toward content pushing meta panel downward once enough exists (as it does in master).

aduth committed:
Tests: Improve block clear provisional verification

Clicking visual editor is not very reliable because clearing is dependant upon where click occurs. Escape is most reliable for intent of clearing the focused selected block.

mtias commented:
Let's get this in, it seems like a good balance of usability and design.
